docs: ECS task shutdown docs for GA (#11508)
hashicorp/consul: website/content/docs/ecs/architecture.mdx

- **T1:** The `sidecar-proxy` container starts. It runs Envoy by executing `envoy -c <path-to-bootstrap-json>`.
- **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, your application containers are started since all Consul machinery is ready to service requests. The only running containers are `consul-client`, `sidecar-proxy`, and your application container(s).

### Task Shutdown

> Graceful shutdown is supported when deploying Consul on ECS, which means the following:

Author: > I'm not sure where the command is issued or if that is even the correct command. I would assume that the user would need to SSH into the VCP?

I think I explained face-to-face, but the ECS tasks will auto join/leave the consul cluster. No manual steps or config required.

> sub-branding a feature called 'Graceful shutdown'.

I agree. For what it's worth, I think it is right to treat this as a feature (we also have "graceful startup" to go along with it). I like the relaxed wording though!

* No incoming traffic from the mesh is directed to this task during shutdown.
* Outgoing traffic to the mesh is possible during shutdown.

This diagram shows an example timeline of a task shutting down:

<img alt="Task Shutdown Timeline" src="/img/ecs-task-shutdown.svg" style={{display: "block", maxWidth: "400px"}} />

- **T0**: ECS sends a TERM signal to all containers.
  - `consul-client` begins to gracefully leave the Consul cluster, since it is configured with `leave_on_terminate = true`
  - `health-sync` stops syncing health status from ECS into Consul checks.

>   - `sidecar-proxy` ignores the TERM signal and continues running until it notices that `user-app` container has exited. This allows the application container to continue to make outgoing requests through the proxy to the mesh. This possible due to an entrypoint override for the container, `consul-ecs envoy-entrypoint`.

Reviewer: ```suggestion
  - `sidecar-proxy` ignores the TERM signal and continues running until the `user-app` container exits. This allows the application container to continue making outgoing requests through the proxy to the mesh. `sidecar-proxy` is able to make the outgoing requests because the `consul-ecs envoy-entrypoint` property is configured for the container. The property overrides the entry point. 
```
I'm not entirely clear on what's happening. This is my interpretation. Is the `consul-ecs envoy-entrypoint` configured by default in such a way that enables this behavior? Is this also configuration information that should be mentioned before describing the process.

>   - `user-app` exits since, in this example, it does not intercept the TERM signal

Reviewer: What example? I thought we were talking in generalities.

- **T1**:
  - `health-sync` updates its Consul checks to critical status, and then exits. This ensures this service instance is marked unhealthy.
  - The `sidecar-proxy` container checks the ECS task metadata. It notices the `user-app` container has stopped, and exits.
- **T2**:
  - `consul-client` finishes leaving the Consul cluster and exits
  - Updates about this task have reached the rest of the Consul cluster, which means downstream proxies are updated to stop sending traffic to this task.
- **T3**: All containers have exited
  - `consul-client` finishes gracefully leaving the Consul datacenter and exits.
  - ECS notices all containers have exited, and will soon put change the Task status to `STOPPED`
- **T4**: (Not applicable to this example, but if any conatiners are still running at this point, ECS forcefully stops them by sending a KILL signal)

> #### Task Shutdown: Completely Avoiding Application Errors

Author: This section is a little long and the examples are kind of out of place in the Architecture page... The content is here for review though, and we can move the example fixes to another page.

> Because Consul service mesh is a distributed, eventually consistent system that is subject to network latency, it is hard to achieve a perfect graceful shutdown.

Reviewer: ```suggestion
Consul service mesh is a distributed and eventually-consistent system subject to network latency, so gracefully shutting down Consul is not always successful.
```
Just wanted to make it sound a little more professional and not like we're saying 'well, sorry, but it's too hard'

> In particular, you may have noticed the following issue in example above, where it is possible that an application that has exited still receives incoming traffic:

Reviewer: ```suggestion
In some cases, an exited application can receive incoming traffic. The following example of this behavior draws on the timeline described above:
```
It wasn't really clear that the process is actually an example. Also, how could the user experience the behavior if we didn't provide any instructions to specifics to walk through the example? It may be less confusing to assume that the user hasn't walked through a tutorial or procedure.

* The `user-app` container exits in **T0**

> * Afterwards in **T2**, downstream services are updated to no longer send traffic to this task

Reviewer: ```suggestion
* During **T2**, downstream services are updated to stop sending traffic to this task.
```

> As a result, downstream applications will see errors when requests are directed to this instance. This can occur for a short period (seconds or less) at the beginning of task shutdown, until the rest of the Consul cluster knows to avoid sending traffic to this instance.

Reviewer: ```suggestion
As a result, downstream applications will report errors when requests are directed to this instance. Errors can be reported for a short period (seconds or less) at the beginning of task shutdown. Applications will stop reporting errors when all nodes in the Consul cluster have instructions to stop sending traffic to this instance.
```
Just trying to use less anthropomorphic language.

> Here are a couple of approaches to address this issue:

Reviewer: ```suggestion
Use the following methods to resolve this issue:
```

1. Modify your application container continue running for a short period of time into task shutdown. By doing this, the application is running to respond to incoming requests successfully at the beginning of task shutdown. This allows time for the Consul cluster to update downstream proxies to stop sending traffic to this task.

>    One way to accomplish this with an entrypoint override for your application container which ignores the TERM signal sent by ECS. Here is an example shell script:

Reviewer: ```suggestion
   You can accomplish this by using an entrypoint override for your application container. Entrypoint overrides ignore the TERM signal sent by ECS. The following shell script contains an entrypoint override:
```

   ```hcl
   module "my_task" {
     source  = "hashicorp/consul-ecs/aws//modules/mesh-task"
     version = "<latest version>"
     container_definitions = [
       {
         name  = "my-app"
         image = "..."
         entryPoint = ["/bin/sh", "-c", file("./app-entrypoint.sh")]
         command = ["python", "manage.py", "runserver", "127.0.0.1:8080"]
         ...
       }
     ...
   }
   ```

   This example sets the `entryPoint` for the container, which overrides the default entrypoint from the image. When the container starts in ECS, the `command` list is passed as arguments to the `entryPoint` command. Putting this together, the container would start with the command, `/bin/sh -c "<app-entrypoint-contents>" python manage.py runserver 127.0.0.1:8080`.

2. If the traffic is HTTP(S), you can enable retry logic through Consul Connect [Service Router](/docs/connect/config-entries/service-router). This will configure proxies retry when receiving an error. When Envoy receives a failed request an upstream service, it can retry the request to a different instance of that service that may be able to respond successfully.

>    To enable retries through Service Router for a service named `example`, first ensure the configured protocol to `http`:

Reviewer 2: ```suggestion
   To enable retries through Service Router for a service named `example`, first ensure the configured protocol is `http`:
```

   ```hcl
   Kind = "service-defaults"
   Name = "example"
   Protocol = "http"
   ```

   The apply the config entry:

   ```shell-session
   $ consul config write example-defaults.hcl
   ```

>    The add retry settings for the service:

Reviewer 2: ```suggestion
   Add retry settings for the service:
```

- **T1:** The `sidecar-proxy` container starts. It runs Envoy by executing `envoy -c <path-to-bootstrap-json>`.
- **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, your application containers are started since all Consul machinery is ready to service requests. The only running containers are `consul-client`, `sidecar-proxy`, and your application container(s).

### Task Shutdown

This diagram shows an example timeline of a task shutting down:

<img alt="Task Shutdown Timeline" src="/img/ecs-task-shutdown.svg" style={{display: "block", maxWidth: "400px"}} />

- **T0**: ECS sends a TERM signal to all containers. Each container reacts to the TERM signal:
  - `consul-client` begins to gracefully leave the Consul cluster.
  - `health-sync` stops syncing health status from ECS into Consul checks.
  - `sidecar-proxy` ignores the TERM signal and continues running until the `user-app` container exits. This allows the application container to continue making outgoing requests through the proxy to the mesh.

>   - `user-app` exits if it does not ignore the TERM signal. Alternatively, if the `user-app` container is configured to ignore the TERM signal , it would continue running.

Reviewer: ```suggestion
  - `user-app` exits if it is not configure to ignore the TERM signal. The `user-app` container will continue running if it is configured to ignore the TERM signal.
```

- **T1**:

>   - `health-sync` updates its Consul checks to critical status, and then exits. This ensures this service instance is marked unhealthy.

Reviewer: ```suggestion
  - `health-sync` updates its Consul checks to critical status and exits. This ensures this service instance is marked unhealthy.
```
